File: src/2015/12/day.12.py
from json import dumps, loads
from timeit import default_timer as timer
from typing import Any

sum = 0


def parse_input():
    pass


def part1():
    global sum

    sum = 0

    with open("src/inputs/2015/12/day.12.input.txt", "r") as file:
        loads(file.read(), parse_int=sum_parse_int)

    return sum


def part2():
    global sum

    sum = 0

    with open("src/inputs/2015/12/day.12.input.txt", "r") as file:
        loads(dumps(loads(
            file.read(), object_hook=ignore_red_object_hook)), parse_int=sum_parse_int)

    return sum


def sum_parse_int(value: str) -> int:
    global sum

    sum += int(value)
    return int(value)


def ignore_red_object_hook(object: dict) -> Any:
    if "red" in object.values():
        return dict()

    return object


def timed(fn):
    start = timer()
    result = fn()

    return (timer() - start, result)


(p1_elapsed, p1_answer) = timed(part1)
print(f"Part 1: {p1_answer} ({p1_elapsed:.3f}s)")

(p2_elapsed, p2_answer) = timed(part2)
print(f"Part 2: {p2_answer} ({p2_elapsed:.3f}s)")

File: src/2015/12/day.12.ts
import fs from "fs";

import { isArray } from "@/utils/array";
import { isNumeric } from "@/utils/number";

function parseInput() {
  return fs.readFileSync("src/inputs/2015/12/day.12.input.txt").toString();
}

export function part1() {
  let total = 0;

  JSON.parse(parseInput(), (_, value) => {
    if (isNumeric(value)) {
      total += Number(value);
    }
  });

  return total;
}

export function part2() {
  let total = 0;

  JSON.stringify(
    JSON.parse(parseInput(), (_, value) => {
      if (!isArray(value) && Object.values(value).includes("red")) {
        return;
      }

      return value;
    }),
    (_, value) => {
      if (isNumeric(value)) {
        total += Number(value);
      }

      return value;
    }
  );

  return total;
}
